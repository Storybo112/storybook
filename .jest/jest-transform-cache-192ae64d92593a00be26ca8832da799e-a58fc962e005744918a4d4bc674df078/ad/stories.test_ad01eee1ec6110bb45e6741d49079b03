'use strict';var _enzyme = require('enzyme');
var _react = require('react');var _react2 = _interopRequireDefault(_react);
var _stories = require('./stories');var _stories2 = _interopRequireDefault(_stories);function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}

describe('manager.ui.components.left_panel.stories', function () {
  describe('render', function () {
    test('should render stories - empty', function () {
      var data = [];
      var wrap = (0, _enzyme.shallow)(_react2.default.createElement(_stories2.default, { stories: data }));

      var list = wrap.find('div').first().children('div').last();

      expect(list.text()).toBe('');
    });

    test('should render stories', function () {
      var data = [{ kind: 'a', stories: ['a1', 'a2'] }, { kind: '20', stories: ['b1', 'b2'] }];
      var wrap = (0, _enzyme.shallow)(_react2.default.createElement(_stories2.default, { stories: data, selectedKind: '20', selectedStory: 'b2' }));

      var output = wrap.html();

      expect(output).toMatch(/20/);
      expect(output).toMatch(/b2/);
    });
  });

  describe('events', function () {
    test('should call the onSelectStory prop when a kind is clicked', function () {
      var data = [{ kind: 'a', stories: ['a1', 'a2'] }, { kind: 'b', stories: ['b1', 'b2'] }];
      var onSelectStory = jest.fn();
      var wrap = (0, _enzyme.shallow)(
      _react2.default.createElement(_stories2.default, {
        stories: data,
        selectedKind: 'b',
        selectedStory: 'b2',
        onSelectStory: onSelectStory }));



      var kind = wrap.find('a').filterWhere(function (el) {return el.text() === 'a';}).last();
      kind.simulate('click');

      expect(onSelectStory).toHaveBeenCalledWith('a', null);
    });

    test('should call the onSelectStory prop when a story is clicked', function () {
      var data = [{ kind: 'a', stories: ['a1', 'a2'] }, { kind: 'b', stories: ['b1', 'b2'] }];
      var onSelectStory = jest.fn();
      var wrap = (0, _enzyme.shallow)(
      _react2.default.createElement(_stories2.default, {
        stories: data,
        selectedKind: 'b',
        selectedStory: 'b2',
        onSelectStory: onSelectStory }));



      var kind = wrap.find('a').filterWhere(function (el) {return el.text() === 'b1';}).last();
      kind.simulate('click');

      expect(onSelectStory).toHaveBeenCalledWith('b', 'b1');
    });
  });
});