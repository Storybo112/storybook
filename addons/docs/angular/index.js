/* eslint-disable no-underscore-dangle */
/* global window */

const setCompodocJson = (compodocJson) => {
  // @ts-ignore
  window.__STORYBOOK_COMPODOC_JSON__ = compodocJson;
};

module.exports.setCompodocJson = setCompodocJson;
