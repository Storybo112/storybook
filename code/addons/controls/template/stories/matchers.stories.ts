import { global as globalThis } from '@storybook/global';
import type { PartialStoryFn, StoryContext } from '@storybook/types';

export default {
  component: globalThis.Components.Pre,
  argTypes: {
    content: {
      control: {
        type: 'nativetextarea',
      },
    },
  },
  decorators: [
    (storyFn: PartialStoryFn, context: StoryContext) =>
      storyFn({ args: { object: { ...context.args } } }),
  ],
};

const Template = ({ content }) => `<pre>${content}</pre>`;

export const Default = Template.bind({});
Default.args = {
  content: 'Hello World',
};
