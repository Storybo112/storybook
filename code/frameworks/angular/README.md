# Storybook for Angular

See [documentation](https://storybook.js.org/docs/angular/get-started/introduction) for installation instructions, usage examples, APIs, and more.
