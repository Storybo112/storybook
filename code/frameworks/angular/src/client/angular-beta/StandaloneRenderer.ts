import { Type, NgModule, Component, OnDestroy, AfterViewInit } from '@angular/core';
import { ApplicationConfig } from '@angular/platform-browser';
import { BehaviorSubject } from 'rxjs';

import { stringify } from 'telejson';

import { StoryFnAngularReturnType, ICollection } from '../types';
import { AbstractRenderer } from './AbstractRenderer';
import { PropertyExtractor } from './utils/PropertyExtractor';
import { getApplication } from './StorybookModule';
import { storyPropsProvider } from './StorybookProvider';
import { getNextStoryUID } from './utils/StoryUID';

export class StandaloneRenderer extends AbstractRenderer {
  protected isFirstRender: boolean = true;

  // constructor(public storyId: string) { }

  public getRenderableComponent({
    storyFnAngular,
    forced,
    // parameters
    component,
    targetDOMNode,
  }: {
    storyFnAngular: StoryFnAngularReturnType;
    forced: boolean;
    // parameters: Parameters;
    component?: any;
    targetDOMNode: HTMLElement;
  }): {
    component: Type<any> | null;
    applicationConfig: ApplicationConfig;
  } | null {
    const targetSelector = this.generateTargetSelectorFromStoryId(targetDOMNode.id);

    const newStoryProps$ = new BehaviorSubject<ICollection | undefined>(storyFnAngular.props ?? {});

    // Storybook now retrieves the component from the context, so this is a workaround to preserve the component in composed stories.
    // eslint-disable-next-line no-underscore-dangle
    const storyComponent = component || (storyFnAngular as any)._composedComponent;

    if (
      !this.fullRendererRequired({
        targetDOMNode,
        storyFnAngular,
        moduleMetadata: {
          ...storyFnAngular.moduleMetadata,
        },
        forced,
      })
    ) {
      this.storyProps$.next(storyFnAngular.props);
      this.isFirstRender = false;
      return null;
    }

    if (!this.isFirstRender) {
      return null;
    }

    this.storyProps$ = newStoryProps$;

    const analyzedMetadata = new PropertyExtractor(
      storyFnAngular.moduleMetadata || {},
      storyComponent
    );

    return {
      component: getApplication({
        storyFnAngular,
        component: storyComponent,
        targetSelector,
        analyzedMetadata,
      }),
      applicationConfig: {
        ...storyFnAngular.applicationConfig,
        providers: [
          storyPropsProvider(newStoryProps$),
          ...(analyzedMetadata.applicationProviders ?? []),
          ...(storyFnAngular.applicationConfig?.providers ?? []),
        ],
      },
    };
  }

  public completeStory(): void {
    // Complete last BehaviorSubject and set a new one for the current module
    if (this.storyProps$) {
      this.storyProps$.complete();
    }
  }

  async beforeFullRender(domNode?: HTMLElement): Promise<void> {}
}

export interface RenderableStoryAndModule {
  component: any;
  applicationConfig: ApplicationConfig;
}

/**
 * Function that will receive a StoryFnAngularReturnType and will return a Component and NgModule that renders the story.
 *
 * @param story
 * @returns
 */
export function createMountable(storyFnReturn: StoryFnAngularReturnType): RenderableStoryAndModule {
  const storyId = `storybook-testing-wrapper`;
  const renderer = new StandaloneRenderer();

  const domNode = document.createElement('span');
  domNode.id = storyId;

  const story: StoryFnAngularReturnType = {
    ...(storyFnReturn as any),
    moduleMetadata: {
      declarations: [...((storyFnReturn as any).moduleMetadata?.declarations ?? [])],
      imports: [...((storyFnReturn as any).moduleMetadata?.imports ?? [])],
      providers: [...((storyFnReturn as any).moduleMetadata?.providers ?? [])],
      entryComponents: [...((storyFnReturn as any).moduleMetadata?.entryComponents ?? [])],
      schemas: [...((storyFnReturn as any).moduleMetadata?.schemas ?? [])],
    },
  };

  const componentAndConfig = renderer.getRenderableComponent({
    storyFnAngular: story,
    forced: false,
    // parameters: {} as any,
    targetDOMNode: domNode,
  });

  const storyUid = getNextStoryUID(storyId);

  // This additional wrapper can probably be avoided by making some changes to
  // the renderer or wrapper component in '@storybook/angular'.
  @Component({
    selector: `sb-testing-mountable[${storyUid}]`,
    template: `<${storyId}></${storyId}>`,
    imports: [(componentAndConfig as any).component],
    standalone: true,
  })
  class SbTestingMountable implements OnDestroy, AfterViewInit {
    ngOnDestroy(): void {
      renderer.completeStory();
    }

    ngAfterViewInit(): void {
      const domNodeInner = document.createElement('span');
      domNodeInner.id = storyId;
      renderer.getRenderableComponent({
        storyFnAngular: storyFnReturn as any,
        forced: false,
        // parameters: {} as any,
        targetDOMNode: domNodeInner,
      });
    }
  }

  if (componentAndConfig === null) {
    throw Error(`Must initially have module`);
  }

  return {
    component: SbTestingMountable,
    applicationConfig: componentAndConfig.applicationConfig,
  };
}
