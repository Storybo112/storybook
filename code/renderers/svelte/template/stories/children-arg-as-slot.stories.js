import ComponentWithSlot from './views/ComponentWithSlot.svelte';
import BorderDecoratorBlue from './views/BorderDecoratorBlue.svelte';

const initialFrameworkOptions = structuredClone(globalThis.FRAMEWORK_OPTIONS);
export default {
  component: ComponentWithSlot,
  beforeEach: () => {
    globalThis.FRAMEWORK_OPTIONS = {
      ...globalThis.FRAMEWORK_OPTIONS,
      childrenArgAsDefaultSlot: true,
    };
    return () => {
      globalThis.FRAMEWORK_OPTIONS = initialFrameworkOptions;
    };
  },
};

export const NoChildren = {};
export const PlainString = {
  args: {
    children: 'This is a plain string',
  },
};
export const HTMLString = {
  args: {
    children: `
      <p>
        This is an HTML string with <i>italic</i>,
        <b>
          bold,
        </b>
        <code>
          some code
        </code>
      </p>
      <p>... and a second paragraph</p>`,
  },
};
export const Number = {
  args: {
    children: 1,
  },
};
export const ComponentWithoutProps = {
  args: {
    children: {
      Component: globalThis.Components.Button,
    },
  },
};
export const ComponentWithProps = {
  args: {
    children: {
      Component: globalThis.Components.Button,
      props: {
        label: 'Button label',
      },
    },
  },
};
export const WithDecorator = {
  decorators: [() => BorderDecoratorBlue],
  args: {
    children: `This is a plain string with a decorator`,
  },
};

export const FrameworkOptionDisabled = {
  beforeEach: () => {
    globalThis.FRAMEWORK_OPTIONS = {
      ...globalThis.FRAMEWORK_OPTIONS,
      childrenArgAsDefaultSlot: false,
    };
    return () => {
      globalThis.FRAMEWORK_OPTIONS = initialFrameworkOptions;
    };
  },
  args: {
    children: `This is just a regular children prop`,
  },
};
