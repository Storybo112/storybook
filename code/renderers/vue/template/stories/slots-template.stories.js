import { global as globalThis } from '@storybook/global';

const MyButton = globalThis.Components.Button;

export default {
  component: MyButton,
  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export const WithDefaultRenderAndDefaultSlot = {
  args: {
    default: 'Button with default render and default slot',
  },
};

export const WithDefaultRenderAndNamedSlot = {
  args: {
    default: 'Button with default render and named slot',
    icon: '<svg width="24" height="24" stroke="currentColor" stroke-width="2" fill="none" stroke-linecap="round" stroke-linejoin="round" class="css-i6dzq1"><path d="m9 18 6-6-6-6"/></svg>',
  },
};
