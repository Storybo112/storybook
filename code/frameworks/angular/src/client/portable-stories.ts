import {
  composeStory as originalComposeStory,
  composeStories as originalComposeStories,
  setProjectAnnotations as originalSetProjectAnnotations,
} from 'storybook/internal/preview-api';
import {
  Args,
  NamedOrDefaultProjectAnnotations,
  StoryAnnotationsOrFn,
  Store_CSFExports,
  StoriesWithPartialProps,
  ProjectAnnotations,
} from 'storybook/internal/types';
import { TestingLibraryMustBeConfiguredError } from 'storybook/internal/preview-errors';
import { TestBed } from '@angular/core/testing';

import { AngularRenderer } from './types';
import { Meta } from './public-types';
import * as angularProjectAnnotations from './entry-preview';
import { createMountable } from './angular-beta/StandaloneRenderer';

/** Function that sets the globalConfig of your storybook. The global config is the preview module of your .storybook folder.
 *
 * It should be run a single time, so that your global config (e.g. decorators) is applied to your stories when using `composeStories` or `composeStory`.
 *
 * Example:
 *```ts
 * // setup.js (for jest)
 * import { setProjectAnnotations } from '@storybook/angular';
 * import * as projectAnnotations from './.storybook/preview';
 *
 * setProjectAnnotations(projectAnnotations);
 *```
 *
 * @param projectAnnotations - e.g. (import * as projectAnnotations from '../.storybook/preview')
 */
export function setProjectAnnotations(
  projectAnnotations:
    | NamedOrDefaultProjectAnnotations<AngularRenderer>
    | NamedOrDefaultProjectAnnotations<AngularRenderer>[]
) {
  originalSetProjectAnnotations<AngularRenderer>(projectAnnotations);
}

// This will not be necessary once we have auto preset loading
export const INTERNAL_DEFAULT_PROJECT_ANNOTATIONS: ProjectAnnotations<AngularRenderer> = {
  ...angularProjectAnnotations,
  renderToCanvas: async ({
    storyFn,
    storyContext: { context, unboundStoryFn: Story, testingLibraryRender: render, canvasElement },
  }) => {
    if (render == null) throw new TestingLibraryMustBeConfiguredError();
    // TODO: Confirm that my reasoning is correct for calling TestBed.resetTestingModule().
    // Since render creates a new TestBed module and Storybook may call render
    // multiple times, we need to manually reset the TestBed module, instead of
    // relying on the afterEach hook.
    TestBed.resetTestingModule();
    const { component, applicationConfig } = createMountable(storyFn({ label: 'Hello world' }));
    const { unmount, container } = await render(component, {
      providers: applicationConfig.providers,
    });

    // I did not see a way to pass the canvasElement to the render function. So,
    // instead render is called then the container is appended to the
    // canvasElement.
    if (canvasElement) {
      canvasElement.appendChild(container);
    }

    return unmount;
  },
};

/**
 * Function that will receive a story along with meta (e.g. a default export from a .stories file)
 * and optionally projectAnnotations e.g. (import * from '../.storybook/preview)
 * and will return a composed component that has all args/parameters/decorators/etc combined and applied to it.
 *
 *
 * It's very useful for reusing a story in scenarios outside of Storybook like unit testing.
 *
 * Example:
 *```ts
 * import { render, screen } from '@testing-library/angular';
 * import {
 *   composeStories,
 *   createMountable,
 * } from '@storybook/testing-angular';
 * import Meta, { Primary as PrimaryStory } from './button.stories';
 *
 * const Primary = composeStory(PrimaryStory, Meta);
 *
 * describe('renders primary button with Hello World', () => {
 *   it('renders primary button with Hello World', async () => {
 *     const { component, applicationConfig } = createMountable(Primary({ label: 'Hello world' }));
 *     await render(component, { providers: applicationConfig.providers });
 *     const buttonElement = screen.getByText(/Hello world/i);
 *     expect(buttonElement).not.toBeNull();
 *   });
 * });
 *```
 *
 * @param story
 * @param componentAnnotations - e.g. (import Meta from './Button.stories')
 * @param [projectAnnotations] - e.g. (import * as projectAnnotations from '../.storybook/preview') this can be applied automatically if you use `setProjectAnnotations` in your setup files.
 * @param [exportsName] - in case your story does not contain a name and you want it to have a name.
 */
export function composeStory<TArgs extends Args = Args>(
  story: StoryAnnotationsOrFn<AngularRenderer, TArgs>,
  componentAnnotations: Meta<TArgs | any>,
  projectAnnotations?: ProjectAnnotations<AngularRenderer>,
  exportsName?: string
) {
  componentAnnotations.decorators = [
    ...(componentAnnotations.decorators ?? ([] as any)),
    (storyInner: any, context: any) => {
      return {
        ...storyInner(context.args),
        // Storybook now retrieves the component from the context, so this is a workaround to preserve the component in composed stories.
        _composedComponent: context.component,
      };
    },
  ];
  return originalComposeStory<AngularRenderer, TArgs>(
    story as StoryAnnotationsOrFn<AngularRenderer, Args>,
    // TODO: Fix types
    componentAnnotations as any,
    projectAnnotations,
    INTERNAL_DEFAULT_PROJECT_ANNOTATIONS,
    exportsName
  );
}

/**
 * Function that will receive a stories import (e.g. `import * as stories from './Button.stories'`)
 * and optionally projectAnnotations (e.g. `import * from '../.storybook/preview`)
 * and will return an object containing all the stories passed, but now as a composed component that has all args/parameters/decorators/etc combined and applied to it.
 *
 *
 * It's very useful for reusing stories in scenarios outside of Storybook like unit testing.
 *
 * Example:
 *```ts
 * import { render, screen } from '@testing-library/angular';
 * import {
 *   composeStory,
 *   createMountable,
 * } from '@storybook/testing-angular';
 * import * as stories from './button.stories';
 * import Meta from './button.stories';
 *
 * const { Primary } = composeStories(stories);
 *
 * describe('button', () => {
 *   it('reuses args from composed story', async () => {
 *     const { component, applicationConfig } = createMountable(Primary({}));
 *     await render(component, { providers: applicationConfig.providers });
 *     expect(screen.getByText(Primary.args?.label!)).not.toBeNull();
 *   });
 * });
 *```
 *
 * @param csfExports - e.g. (import * as stories from './Button.stories')
 * @param [projectAnnotations] - e.g. (import * as projectAnnotations from '../.storybook/preview') this can be applied automatically if you use `setProjectAnnotations` in your setup files.
 */
export function composeStories<TModule extends Store_CSFExports<AngularRenderer, any>>(
  csfExports: TModule,
  projectAnnotations?: ProjectAnnotations<AngularRenderer>
) {
  const composedStories = originalComposeStories(csfExports, projectAnnotations, composeStory);

  return composedStories as unknown as Omit<
    StoriesWithPartialProps<AngularRenderer, TModule>,
    keyof Store_CSFExports
  >;
}
