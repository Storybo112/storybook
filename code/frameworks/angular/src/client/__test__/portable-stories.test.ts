// @vitest-environment happy-dom

/// <reference types="@testing-library/jest-dom" />;
import { it, expect, vi, describe, afterEach } from 'vitest';
import { render, screen } from '@testing-library/angular';
import { expectTypeOf } from 'expect-type';
import { Meta } from '..';
import * as stories from './button.stories';
import { ButtonComponent } from './button.component';
import { composeStories, composeStory, setProjectAnnotations } from '../portable-stories';
import { createMountable } from '../angular-beta/StandaloneRenderer';

setProjectAnnotations({ testingLibraryRender: render as any });

// example with composeStories, returns an object with all stories composed with args/decorators
const { CSF3Primary, LoaderStory } = composeStories(stories);

// example with composeStory, returns a single story composed with args/decorators
const Secondary = composeStory(stories.CSF2Secondary, stories.default);
describe('renders', () => {
  it('renders primary button with custom props via composeStory', async () => {
    // We unfortunately can't do the following:
    // render(CSF3Primary.Component, { ...CSF3Primary.props, label: 'Hello world' });
    // Because the props will be passed to the first decorator of the story instead
    // of the actual component of the story. This is because of our current PreviewRender structure

    const Composed = composeStory(
      {
        ...stories.CSF3Primary,
        args: { ...stories.CSF3Primary.args, label: 'Hello world' },
      },
      stories.default
    );

    const { component, applicationConfig } = createMountable(Composed());
    await render(component, { providers: applicationConfig.providers });
    const buttonElement = await screen.getByText(/Hello world/i);
    await expect(buttonElement).not.toBeNull();
  });

  it('reuses args from composed story', async () => {
    const { component, applicationConfig } = createMountable(Secondary());
    await render(component, { providers: applicationConfig.providers });
    const buttonElement = screen.getByRole('button');
    await expect(buttonElement.textContent).toMatch(Secondary.args.label);
  });

  it('onclick handler is called', async () => {
    const onClickSpy = vi.fn();
    const { component, applicationConfig } = createMountable(Secondary({ onClick: onClickSpy }));
    await render(component, { providers: applicationConfig.providers });
    const buttonElement = screen.getByRole('button');
    buttonElement.click();
    expect(onClickSpy).toHaveBeenCalled();
  });

  it('reuses args from composeStories', async () => {
    const { component, applicationConfig } = createMountable(CSF3Primary());
    const { getByText } = await render(component, { providers: applicationConfig.providers });
    const buttonElement = getByText(/foo/i);
    expect(buttonElement).not.toBeNull();
  });

  it('should call and compose loaders data', async () => {
    await LoaderStory.load();
    const { component, applicationConfig } = createMountable(LoaderStory());
    const { getByTestId } = await render(component, { providers: applicationConfig.providers });
    await expect(getByTestId('spy-data').textContent).toEqual('mockFn return value');
    await expect(getByTestId('loaded-data').textContent).toEqual('loaded data');
    // spy assertions happen in the play function and should work
    await LoaderStory.play!();
  });
});

// describe('projectAnnotations', () => {
//   it('renders with default projectAnnotations', async () => {
//     setProjectAnnotations([
//       {
//         parameters: { injected: true },
//         globalTypes: {
//           locale: { defaultValue: 'en' },
//         },
//         testingLibraryRender: render as any,
//       },
//     ]);
//     const WithEnglishText = composeStory(stories.CSF2StoryWithLocale, stories.default);
//     const { component, applicationConfig } = createMountable(WithEnglishText());
//     const { getByText } = await render(component, { providers: applicationConfig.providers });
//     const buttonElement = getByText('Hello!');
//     expect(buttonElement).not.toBeNull();
//     expect(WithEnglishText.parameters?.injected).toBe(true);
//   });

//   it('renders with custom projectAnnotations via composeStory params', async () => {
//     const WithPortugueseText = composeStory(stories.CSF2StoryWithLocale, stories.default, {
//       initialGlobals: { locale: 'pt' },
//     });
//     const { component, applicationConfig } = createMountable(WithPortugueseText());
//     const { getByText } = await render(component, { providers: applicationConfig.providers });
//     const buttonElement = getByText('OlÃ¡!');
//     expect(buttonElement).not.toBeNull();
//   });
// });

describe('CSF3', () => {
  it('renders with inferred globalRender', async () => {
    const Primary = composeStory(stories.CSF3Button, stories.default);
    const { component, applicationConfig } = createMountable(Primary());
    await render(component, { providers: applicationConfig.providers });
    const buttonElement = screen.getByText(/foo/i);
    expect(buttonElement).not.toBeNull();
  });

  it('renders with custom render function', async () => {
    const Primary = composeStory(stories.CSF3ButtonWithRender, stories.default);

    const { component, applicationConfig } = createMountable(Primary());
    await render(component, { providers: applicationConfig.providers });
    expect(screen.getByTestId('custom-render')).not.toBeNull();
  });

  it('renders with play function without canvas element', async () => {
    const CSF3InputFieldFilled = composeStory(stories.CSF3InputFieldFilled, stories.default);

    await CSF3InputFieldFilled.play();

    const input = screen.getByTestId('input') as HTMLInputElement;
    expect(input.value).toEqual('Hello world!');
  });

  it('renders with play function with canvas element', async () => {
    const CSF3InputFieldFilled = composeStory(stories.CSF3InputFieldFilled, stories.default);

    const div = document.createElement('div');
    document.body.appendChild(div);

    await CSF3InputFieldFilled.play({ canvasElement: div });

    const input = screen.getByTestId('input') as HTMLInputElement;
    expect(input.value).toEqual('Hello world!');

    document.body.removeChild(div);
  });
});

describe('ComposeStories types', () => {
  // this file tests Typescript types that's why there are no assertions
  it('Should support typescript operators', () => {
    type ComposeStoriesParam = Parameters<typeof composeStories>[0];

    expectTypeOf({
      ...stories,
      default: stories.default as Meta<ButtonComponent>,
    }).toMatchTypeOf<ComposeStoriesParam>();

    expectTypeOf({
      ...stories,

      /**
       * Types of property 'argTypes' are incompatible.
       * Type '{ backgroundColor: { control: string; }; size: { control: { type: string; }; options: string[]; }; }'
       * has no properties in common with type 'Partial<ArgTypes<ComponentType>>'.
       */
      default: stories.default satisfies Meta<ButtonComponent>,
    }).toMatchTypeOf<ComposeStoriesParam>();
  });
});

// Batch snapshot testing
const testCases = Object.values(composeStories(stories)).map(
  (Story) => [Story.storyName, Story] as [string, typeof Story]
);
it.each(testCases)('Renders %s story', async (_storyName, Story) => {
  if (_storyName === 'CSF2StoryWithLocale') return;
  await Story.play();
  expect(document.body).toMatchSnapshot();
});
