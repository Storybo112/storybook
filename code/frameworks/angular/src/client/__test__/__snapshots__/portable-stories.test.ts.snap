// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders CSF2Secondary story 1`] = `
<body>
  <div
    id="root9"
    ng-version="17.3.0"
  >
    <storybook-testing-wrapper>
      <storybook-button
        ng-reflect-label="label coming from story args!"
        ng-reflect-primary="false"
      >
        <button
          class="storybook-button storybook-button--medium storybook-button--secondary"
          type="button"
        >
           label coming from story args! 
        </button>
      </storybook-button>
      <!--container-->
    </storybook-testing-wrapper>
  </div>
</body>
`;

exports[`Renders CSF2StoryWithParamsAndDecorator story 1`] = `
<body>
  <div
    id="root10"
    ng-version="17.3.0"
  >
    <storybook-testing-wrapper>
      <div
        style="margin: 3em;"
      >
        undefined
      </div>
    </storybook-testing-wrapper>
  </div>
</body>
`;

exports[`Renders CSF3Button story 1`] = `
<body>
  <div
    id="root12"
    ng-version="17.3.0"
  >
    <storybook-testing-wrapper>
      <storybook-button
        ng-reflect-label="foo"
        ng-reflect-primary="false"
      >
        <button
          class="storybook-button storybook-button--medium storybook-button--secondary"
          type="button"
        >
           foo 
        </button>
      </storybook-button>
      <!--container-->
    </storybook-testing-wrapper>
  </div>
</body>
`;

exports[`Renders CSF3ButtonWithRender story 1`] = `
<body>
  <div
    id="root13"
    ng-version="17.3.0"
  >
    <storybook-testing-wrapper>
      <div>
        <p
          data-testid="custom-render"
        >
          I am a custom render function
        </p>
         undefined 
      </div>
    </storybook-testing-wrapper>
  </div>
</body>
`;

exports[`Renders CSF3InputFieldFilled story 1`] = `
<body>
  <div
    id="root14"
    ng-version="17.3.0"
  >
    <storybook-testing-wrapper>
      <input
        data-testid="input"
      />
    </storybook-testing-wrapper>
  </div>
</body>
`;

exports[`Renders CSF3Primary story 1`] = `
<body>
  <div
    id="root11"
    ng-version="17.3.0"
  >
    <storybook-testing-wrapper>
      <storybook-button
        ng-reflect-label="foo"
        ng-reflect-primary="true"
        ng-reflect-size="large"
      >
        <button
          class="storybook-button storybook-button--large storybook-button--primary"
          type="button"
        >
           foo 
        </button>
      </storybook-button>
      <!--container-->
    </storybook-testing-wrapper>
  </div>
</body>
`;

exports[`Renders LoaderStory story 1`] = `
<body>
  <div
    id="root15"
    ng-version="17.3.0"
  >
    <storybook-testing-wrapper>
      <div>
        <div
          data-testid="loaded-data"
        >
          loaded data
        </div>
        <div
          data-testid="spy-data"
        >
          mockFn return value
        </div>
      </div>
    </storybook-testing-wrapper>
  </div>
</body>
`;
